deliaioanac/assignment3/Dictionary.cpp

```diff
@@ -0,0 +1,89 @@
+#include <iostream>
```

Comment:
[optional] Trie is the optimal implementation for a dictionary, however you're not taking full advantage of the data structure :)

For each `backtrack` call you look up prefix/word from the root, which is inefficient, up to O(N) when you look up the whole word. You can improve this lookup time to O(1), by passing around the current Trie node, such that you only have to preform lookups from a subtree you haven't explored yet, if that makes sense.

Reply:
I am not sure how to implement that without taking the root_ member of the dictionary from private to public, which would expose the internal implementation of the Trie. It would be more efficient, but I am not sure how to do that while still maintaining good abstraction.

Reply:
Yeah, that's a good point. 

Typically if you're implementing a generic implementation of a data structure for a common library you don't want to return anything that would indicate the underlying implementation as it would be a nightmare to change if at some point down the line you decide that a Trie is better implemented with `LinkedLists` rather than `TrieNodes` :)

You have two options here:

1. Don't make 'generic' solutions for specific problems. This is perfectly fine in software engineering, and is often preferred in agile-like programming where you want to get solutions out fast, and iterate on them later. In my daily work, I often do this :) If I need a generic solution, there's a std library that already does that. If I'm ever implementing a BinaryTree myself it's because the std library solution is too generic, so there's no point in trying to keep my implementation generic too. In your case you could do something like this:

```
// Returns nullptr if the string is not a valid prefix, otherwise returns the Trie pointer to the last
// character of the prefix.
const TrieNode* const isPrefix(...) {...}

```

I think it's `const * const`, but you'd have to double check :) I always get these wrong, but basically you can return the underlying implementation pointer that's not modifiable, so it's slightly better than making `root_` public.

Then in `isPrefixHelper`, you'll need to return the raw pointer to the node with `root->get()`. Note that this is also OK. The `Trie` class still maintains ownership of the node through the unique_ptr and the function that gets the raw pointer cannot modify what it's pointing to or the data it's pointing to :) 

2. You can follow the [Fundamental Theorem of Software Engineering](https://en.wikipedia.org/wiki/Fundamental_theorem_of_software_engineering) and hide everything behind interfaces :) This is a huge overkill for this type of problem, but you could define a `Trie` interface that has a `isPrefix` method that returns a `TriePosition` interface. Then Grid is only reliant on the interface definitions and you're free to keep your implementation details private. 

Interfacing everything is very useful when it comes to tests. If your classes rely on interfaces, you can easily 'mock' out dependency classes and only test whether your own class works.